source/http/HttpConnectionManager.cpp

```diff
@@ -0,0 +1,294 @@
+                 */
+                m_connectionsLock.lock();
+
+                /* Case 1 */
+                if (!m_connections.empty())
+                {
+                    auto connection = m_connections.back();
+                    m_connections.pop_back();
+                    ++m_outstandingVendedConnections;
+                    m_connectionsLock.unlock();
+                    onClientConnectionAvailable(connection, AWS_ERROR_SUCCESS);
+                    return true;
+                }
+
+                /* case 2 or 3, we don't know yet. */
+                m_pendingConnectionRequests.push_back(onClientConnectionAvailable);
+
+                /* if we have available space, case 2, otherwise case 3 */
+                if (!s_createConnection())
+                {
+                    /* case 2 and the connection creation failed, pop back, because in this rare case, we want to just
+                     * tell the caller that our entire attempt at this failed. */
+                    m_pendingConnectionRequests.pop_back();
+                    m_lastError = aws_last_error();
+                    m_connectionsLock.unlock();
+                    onClientConnectionAvailable(nullptr, m_lastError);
+                    return false;
+                }
+
+                m_connectionsLock.unlock();
+                return true;
+            }
+
+            /*
+             * User is finished with a connection and wants to return it to the pool.
+             *
+             * Case 1, The connection is actually dead and can't be returned to the pool.
+             *      Case 1.1 We don't have any pending connection acquisition requests so just let it hit the floor, it
+             * will get created in the next Acquire call. Case 1.2 We have pending connection acquisition requests, so
+             * create a new connection to replace it, let the callback's fire to notify the user and grow the pool as
+             * normal.
+             *
+             * Case 2, the connection is still good.
+             *      Case 2.1 We have pending acquisitions. Just give it to the top of the queue.
+             *      Case 2.2 We don't have pending acquisitions, push it to the pool.
+             */
+            void HttpClientConnectionManager::s_poolOrVendConnection(
+                std::shared_ptr<HttpClientConnection> connection,
+                bool isRelease) noexcept
+            {
+                /** Locking logic:
+                 *  We can't hold the lock during the callback since a user may call back into the connection manager
+                 *  from the callbacks. We don't use a scoped lock because we need finer grain control.
+                 *  All unlocks, should either be followed immediately by a callback and then a return, or
+                 *  a return only. If you see anything else, it's a bug.
+                 */
+                m_connectionsLock.lock();
```

Comment:
Another way you can handle this locking situation (to improve clarity and reduce peril for maintenance) is to use the Synchronized trick (you can find examples in Abseil and Folly);

Basically, associate the mutex with a structure that you want to modify under the mutex's protection.

```cpp
template <typename T, typename Mutex>
class Synchronized
{
public:
    void Modify(std::function<void(T&)> modify) {
        std::lock_guard<Mutex> lock(&m_mutex);
        modify(m_t);
    }
private:
    T m_t;
    Mutex m_mutex;
}
```
And then stick the protected data in a structure like so:
```cpp
struct {
    m_connections;
    m_pendingConnectionRequests;
    m_pendingConnections;
    m_outstandingVendedConnections
}
```

And then pass this logic in as lambdas to Modify